### ISSUE TITLE
Is the tcp sent/received calculated correctly?

### ISSUE THREAD

**Issue by fyuan1316** [kiali/kiali#6760]
First of all, thank you for developing such great software.
I don't quite understand the tcp send and received calculations in the graph api. I would like to ask for your help.

- test case
I am using the sleep and tcp-echo services from the istio sample as workloads for testing. The sleep deployment exists in fy-ns1 namespace and tcp-echo in fy-ns2. I then executed the `echo xxx | nc tcp-echo.fy-ns2` command in the sleep container and observed the captured tcp traffic through prometheus.

- result
I got the following metrics:
> istio_tcp_received_bytes_total
![image](https://github.com/kiali/kiali/assets/42265657/7ea117d8-30db-4f32-bc22-cea735e9fa86)

> istio_tcp_sent_bytes_total
![image](https://github.com/kiali/kiali/assets/42265657/5b00d524-138e-4edf-9c73-7df5a699f20d)

*Combining the experimental and metrics results, I have the following understanding:*
|reporter|source_workload|destination_workload|bytes|desc|metrcis|
|-|-|-|-|-|-|
|source|sleep|tcp-echo|9| sleep sent|istio_tcp_received_bytes_total|
|destination|sleep|tcp-echo|27| tcp-echo sent|istio_tcp_sent_bytes_total|
|destination|sleep|tcp-echo|9| tcp-echo recv| istio_tcp_received_bytes_total|
|source|sleep|tcp-echo|27| sleep recv|istio_tcp_sent_bytes_total|

In the code implementation when I select tcp send or received as the query condition, the kiali log output is as follows:
```
#########
# tcp sent

2023-10-22T11:54:03+08:00 TRC Graph query:

round(sum(rate(istio_tcp_received_bytes_total{reporter="source",source_workload_namespace!="fy",destination_workload_namespace="unknown",destination_workload="unknown",destination_service=~"^.+\\.fy\\..+$"} [300s])) by (source_cluster,source_workload_namespace,source_workload,source_canonical_service,source_canonical_revision,destination_cluster,destination_service_namespace,destination_service,destination_service_name,destination_workload_namespace,destination_workload,destination_canonical_service,destination_canonical_revision,response_flags) > 0,0.001)@time=2023-10-22 11:54:03 (now=2023-10-22 11:54:03, 1697946843)

  
2023-10-22T11:54:03+08:00 TRC Graph query:

round(sum(rate(istio_tcp_received_bytes_total{reporter="destination",destination_workload_namespace="fy"} [300s])) by (source_cluster,source_workload_namespace,source_workload,source_canonical_service,source_canonical_revision,destination_cluster,destination_service_namespace,destination_service,destination_service_name,destination_workload_namespace,destination_workload,destination_canonical_service,destination_canonical_revision,response_flags) > 0,0.001)@time=2023-10-22 11:54:03 (now=2023-10-22 11:54:03, 1697946843)


2023-10-22T11:54:03+08:00 TRC Graph query:

round(sum(rate(istio_tcp_received_bytes_total{reporter="source",source_workload_namespace="fy"} [300s])) by (source_cluster,source_workload_namespace,source_workload,source_canonical_service,source_canonical_revision,destination_cluster,destination_service_namespace,destination_service,destination_service_name,destination_workload_namespace,destination_workload,destination_canonical_service,destination_canonical_revision,response_flags) > 0,0.001)@time=2023-10-22 11:54:03 (now=2023-10-22 11:54:03, 1697946843)

###########
#tcp received


2023-10-22T11:57:36+08:00 TRC Graph query:

round(sum(rate(istio_tcp_sent_bytes_total{reporter="source",source_workload_namespace!="fy",destination_workload_namespace="unknown",destination_workload="unknown",destination_service=~"^.+\\.fy\\..+$"} [300s])) by (source_cluster,source_workload_namespace,source_workload,source_canonical_service,source_canonical_revision,destination_cluster,destination_service_namespace,destination_service,destination_service_name,destination_workload_namespace,destination_workload,destination_canonical_service,destination_canonical_revision,response_flags) > 0,0.001)@time=2023-10-22 11:57:36 (now=2023-10-22 11:57:36, 1697947056)


2023-10-22T11:57:36+08:00 TRC Graph query:

round(sum(rate(istio_tcp_sent_bytes_total{reporter="destination",destination_workload_namespace="fy"} [300s])) by (source_cluster,source_workload_namespace,source_workload,source_canonical_service,source_canonical_revision,destination_cluster,destination_service_namespace,destination_service,destination_service_name,destination_workload_namespace,destination_workload,destination_canonical_service,destination_canonical_revision,response_flags) > 0,0.001)@time=2023-10-22 11:57:36 (now=2023-10-22 11:57:36, 1697947056)


2023-10-22T11:57:36+08:00 TRC Graph query:

round(sum(rate(istio_tcp_sent_bytes_total{reporter="source",source_workload_namespace="fy"} [300s])) by (source_cluster,source_workload_namespace,source_workload,source_canonical_service,source_canonical_revision,destination_cluster,destination_service_namespace,destination_service,destination_service_name,destination_workload_namespace,destination_workload,destination_canonical_service,destination_canonical_revision,response_flags) > 0,0.001)@time=2023-10-22 11:57:36 (now=2023-10-22 11:57:36, 1697947056)
```

*To summarize, in the code implementation(Ignore the unserviced destination for now)*

|direction|reporter|metrics|desc|
|-|-|-|-|
|sent|destination|istio_tcp_received_bytes_total|destination sent ? should we use`istio_tcp_sent_bytes_total `?|
|sent|source|istio_tcp_received_bytes_total|source sent|
|received|destination|istio_tcp_sent_bytes_total|destination received ? should we use `istio_tcp_received_bytes_total `?|
|received|source|istio_tcp_sent_bytes_total|source received|

So,is this a bug, or am I not understanding it correctly? I hope to get your help.

**Comment by jshaughn** [comment]
Hi @fyuan1316, excellent analysis.   You have found an area that is up to some interpretation.  In fact, for years we performed the queries in the way that you describe.  But recently, about 3 months ago, we reversed things by essentially swapping `istio_tcp_sent_bytes_total` and `istio_tcp_received_bytes_total` in these queries.   Even the Istio maintainers are in some disagreement.  In the end, we decided Kiali should be consistent with the grafana dashboards supplied with Istio.  If you look at those dashboards and also at Kiali, you should see similar reporting. Please read through these two issues and the associated PR for more background,  and maybe also look at the Istio dashboards.   And then let us know what you think...

https://github.com/istio/istio/issues/32399
https://github.com/kiali/kiali/issues/5822
https://github.com/kiali/kiali/pull/6323

**Comment by fyuan1316** [comment]
Thanks for your quick reply. 

Considering If the istio community is not going to change the way the metrics are generated, in order to ensure the accuracy of the data, I think it may be necessary to consider swapping metrics without changing the direction of the arrows in the graph to solve this problem. Like you said in #5822.

**Comment by jshaughn** [comment]
Istio is not going to make any changes to the telemetry.  Their answer was to swap the metrics in their dashboards.  That is why we swapped the metrics in our graph generation.  We have already made the change (since v1.71).  So, your analysis is good, but Kiali does not plan to make another change (back to the way it was). As far as I know, we now use the same approach as  the throughput reporting used by the istio dashboards.

**Comment by jshaughn** [comment]
Hi @fyuan1316 , are you satisfied with the explanation above?  If you still believe there is a Kiali bug let us know, otherwise I think this issue can be closed.

**Comment by fyuan1316** [comment]
I don't know if this is a bug. just don't quite understand the significance of the calculations. As I said above. When the sent traffic is selected for the query, the actual calculation is 9 bytes+9bytes, which seems to double count the traffic on the sender and receiver sides of a one-way send.
If that's how it's designed, I think we can close this issue.

**Comment by jshaughn** [comment]
Adding to the backlog to take a closer look, and verify if the rates make sense or are doubled.

**Comment by leandroberetta** [comment]
> _To summarize, in the code implementation(Ignore the unserviced destination for now)_
> 
> direction	reporter	metrics	desc
> sent	destination	istio_tcp_received_bytes_total	destination sent ? should we use`istio_tcp_sent_bytes_total `?
> sent	source	istio_tcp_received_bytes_total	source sent
> received	destination	istio_tcp_sent_bytes_total	destination received ? should we use `istio_tcp_received_bytes_total `?
> received	source	istio_tcp_sent_bytes_total	source received
> So,is this a bug, or am I not understanding it correctly? I hope to get your help.

Hi @fyuan1316, I've looked into this issue and as we mentioned before, given that Istio also does calculations this way (see https://github.com/istio/istio/issues/32399), I can confirm metrics are ok in terms of which ones we are using to calculate. 



> I don't know if this is a bug. just don't quite understand the significance of the calculations. As I said above. When the sent traffic is selected for the query, the actual calculation is 9 bytes+9bytes, which seems to double count the traffic on the sender and receiver sides of a one-way send. If that's how it's designed, I think we can close this issue.

However I couldn't reproduce what you're mentioning here, about doubling the calculations, can you share more about how you're determine that?. Are you doing query into Prometheus based on the queries Kiali's log is logging?.

**Comment by leandroberetta** [comment]
Hi @fyuan1316, provided the information above, are you still having problems with tcp metrics?. As I commented, I couldn't reproduce the calculations you've done, let us know if you still have issues. Thanks.

**Comment by fyuan1316** [comment]
Since this issue is no longer blocking me, I am closing it. Thank you.
